### PR TITLE
feat(porep): Interactive PoRep

### PR DESCRIPTION
**Pull request by laser**
## Why does this PR exist?

Somebody went and changed seal from a one to two-step process.

## What's in this PR?

This PR updates the sector builder to conform to the new sealing protocol.

### CODE REVIEW

File: sector-builder-ffi/examples/simple/plumbing.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ pub(crate) unsafe fn generate_post(
     ptr: *mut sector_builder_ffi_SectorBuilder,
     challenge_seed: [u8; 32],
     proving_set: &ProvingSet,
-) -> Vec<u8> {
+) -> std::result::Result<Vec<u8>, sector_builder_ffi_FCPResponseStatus> {
```
Comment (dignifiedquire) [inline sector-builder-ffi/examples/simple/plumbing.rs]: why the `std::result::Result` prefix?

Comment (dignifiedquire) [reply sector-builder-ffi/examples/simple/plumbing.rs]: given that these all have the same Error type, would be nicer to define a type alias, like `FfiResult`

File: sector-builder-ffi/examples/simple/porcelain.rs [hunk 2]
```diff
@@ -88,10 +100,42 @@ pub(crate) unsafe fn resume_seal_sector_nonblocking(
...
+    ptr: *mut sector_builder_ffi_SectorBuilder,
+    sector_id: u64,
+    seed: sector_builder_ffi_FFISealSeed,
+) {
+    let atomic_ptr = AtomicPtr::new(ptr);
+
+    thread::spawn(move || {
+        let sector_builder = atomic_ptr.into_inner();
+
+        let _ = seal_commit(&mut Default::default(), sector_builder, sector_id, seed);
```
Comment (dignifiedquire) [inline sector-builder-ffi/examples/simple/porcelain.rs]: don't you want to log the error at least somewhere? or is that caught somewhere else?

File: sector-builder/src/helpers/add_piece.rs [hunk 3]
```diff
@@ -39,41 +38,82 @@ pub fn add_piece(
...
+        .get_mut(&dest_sector_id)
+        .ok_or_else(|| format_err!("unable to retrieve sector from state-map"))?;
+
+    // TODO: Buffering the piece completely into memory is awful, but each of
+    // the two function calls (add_piece and generate_piece_commitment) accept a
+    // Read. Given that the piece byte-stream is represented by a Read, we can't
+    // read all of its bytes in one function call and then do the same with the
+    // other w/out putting the bytes into an intermediate buffer. A tee reader
+    // would be appropriate here.
+    let mut backing_buffer = vec![];
```
Comment (dignifiedquire) [inline sector-builder/src/helpers/add_piece.rs]: you could create it with a capacity of the pice size right?

File: sector-builder/src/helpers/add_piece.rs [hunk 4]
```diff
@@ -39,41 +38,82 @@ pub fn add_piece(
...
+    let mut cursor = Cursor::new(&mut backing_buffer);
+
+    std::io::copy(&mut piece_file, &mut cursor)
+        .map_err(|err| format_err!("unable to copy piece bytes to buffer: {:?}", err))?;
```
Comment (dignifiedquire) [inline sector-builder/src/helpers/add_piece.rs]: (no need to change, just an fyi)

No need to a cursor for this first part, 

```rust
let mut buffer = Vec::new();
piece_file.read_to_end(&mut buffer)?;

// now create a cursor for passing it to add_piece
let cursor = Cursor::new(&mut buffer);
```

File: sector-builder/src/metadata_manager.rs [hunk 5]
```diff
@@ -261,67 +293,130 @@ impl<T: KeyValueStore> SectorMetadataManager<T> {
...
-    // Create a SealTaskPrototype for each staged sector matching the predicate.
-    // If a ticket is already associated with the staged sector, use it to
-    // create the proto. Otherwise use the provided ticket.
-    pub fn create_seal_task_protos<P: FnMut(&StagedSectorMetadata) -> bool>(
-        &mut self,
-        seal_ticket: SealTicket,
-        predicate: P,
-    ) -> Result<Vec<SealTaskPrototype>> {
-        let mut protos: Vec<SealTaskPrototype> = Default::default();
+        let (meta, ticket) = match (mode, meta.seal_status.clone()) {
```
Comment (dignifiedquire) [inline sector-builder/src/metadata_manager.rs]: why do you need to always `clone`?

File: sector-builder-ffi/examples/simple/plumbing.rs (hunk 1 above)
Comment (laser) [reply sector-builder-ffi/examples/simple/plumbing.rs]: fixed

File: sector-builder-ffi/examples/simple/porcelain.rs (hunk 2 above)
Comment (laser) [reply sector-builder-ffi/examples/simple/porcelain.rs]: fixed

File: sector-builder/src/metadata_manager.rs (hunk 5 above)
Comment (laser) [reply sector-builder/src/metadata_manager.rs]: fixed

File: sector-builder/src/helpers/add_piece.rs (hunk 4 above)
Comment (laser) [reply sector-builder/src/helpers/add_piece.rs]: word. good tip. in the name of moving forward, i'll just replace this Cursor stuff with code that consumes your new API (in the next PR)

**Pull request merged**